### PR TITLE
[jit] Emit SQRTSD instead of FSQRT on AMD64 and use the libc sin/cos functions.

### PR DESCRIPTION
**Pull request by vkargov**
This patch replaces the "old" x87 division that we currently emit for Math.Sqrt with a "new" SSE one.

The main reason for this is speed, since aside from longer execution time, one such FPU instruction amidst SSE code requires data conversion to/from the 80-bit format, as well.

CoreCLR uses SQRTSD, [too](https://github.com/dotnet/coreclr/blob/master/src/jit/instr.cpp#L3560).

Our benchmarking suite has [at least one test](https://github.com/xamarin/benchmarker/tree/master/tests/csgrande/GrandeTracer) where verification starts failing after this change. However, this is a problem with the test expecting too much (more than the standard requires) precision from the computations, and it's been failing on OS X for quite some time, too, due to similar overly strict precision requirements to another math function in libc.

### CODE REVIEW

**Comment by evincarofautumn** [comment]
I think the reduced precision is acceptable if we can avoid the x87 stack and we match CoreCLR.

Here’s an idea for a small related optimisation: special-case `Math.Pow (x, y)` for `y` = ±0.5 to compute the (inverse) square root when `x` ≠ `-0.0`.

File: mono/mini/mini-amd64.c [hunk 1]
```diff
@@ -5349,7 +5349,7 @@ mono_arch_output_basic_block (MonoCompile *cfg, MonoBasicBlock *bb)
 			break;		
 		}
 		case OP_SQRT:
-			EMIT_SSE2_FPFUNC (code, fsqrt, ins->dreg, ins->sreg1);
+			amd64_sse_sqrtsd_reg_reg (code, ins->dreg, ins->sreg1);
```
Comment (kumpera) [inline mono/mini/mini-amd64.c]: You should adjust cpu-amd64.md as well given the max size changed.

**Comment by kumpera** [comment]
The precision change is fine

File: mono/mini/mini-amd64.c (hunk 1 above)
Comment (vkargov) [inline mono/mini/mini-amd64.c]: Fixed that.

File: mono/mini/mini-amd64.c [hunk 2]
```diff
@@ -8176,11 +8176,15 @@ mono_arch_emit_inst_for_method (MonoCompile *cfg, MonoMethod *cmethod, MonoMetho
 	int opcode = 0;
 
 	if (cmethod->klass == mono_defaults.math_class) {
+#if 0
+		/* Mono can emit Sin and Cos directly as fsin and fcos, but modern libc implementations should be faster. */
```
Comment (evincarofautumn) [inline mono/mini/mini-amd64.c]: This also avoids precision/range problems with `fsin` and `fcos`. :)

**Comment by vkargov** [comment]
I probably should have made another PR for that, but I also made Mono use the libc versions of sin/cos instead of the hardware fsin/fcos versions.
CoreCLR does that, [too](https://github.com/dotnet/coreclr/blob/08786f20e89eb5f518d8d25f3e7f886f69d994ea/src/classlibnative/float/floatdouble.cpp#L229), and it is faster (1.5-3x speedup by my measurements), since all major libc libraries seem to provide their own SSE implementations that don't rely on the x87 instructions.

File: mono/mini/mini-amd64.c (hunk 2 above)
Comment (kumpera) [inline mono/mini/mini-amd64.c]: This is not the way to do it. The code should be removed including on the backends.

How does this impact arm?

**Pull request closed without merging**